includes/embeds/class-amp-tumblr-embed-handler.php

```diff
@@ -6,56 +6,116 @@
 class AMP_Tumblr_Embed_Handler extends AMP_Base_Embed_Handler {
 
+	/**
+	 * Default width.
+	 *
+	 * Tumblr embeds for web have a fixed width of 540px.
+	 *
+	 * @link https://tumblr.zendesk.com/hc/en-us/articles/226261028-Embed-pro-tips
+	 * @var int
+	 */
+	protected $DEFAULT_WIDTH = 540;
+
+	/**
+	 * Base URL used for identifying embeds.
+	 *
+	 * @var string
+	 */
+	protected $base_embed_url = 'https://embed.tumblr.com/embed/post/';
+
 	/**
 	 * Register embed.
 	 */
 	public function register_embed() {
-		add_filter( 'embed_oembed_html', [ $this, 'filter_embed_oembed_html' ], 10, 2 );
+		// Not implemented.
 	}
 
 	/**
 	 * Unregister embed.
 	 */
 	public function unregister_embed() {
-		remove_filter( 'embed_oembed_html', [ $this, 'filter_embed_oembed_html' ], 10 );
+		// Not implemented.
 	}
 
 	/**
-	 * Filter oEmbed HTML for Tumblr to prepare it for AMP.
+	 * Sanitizes Tumblr raw embeds to make them AMP compatible.
 	 *
-	 * @param string $cache Cache for oEmbed.
-	 * @param string $url   Embed URL.
-	 * @return string Embed.
+	 * @param Document $dom DOM.
 	 */
-	public function filter_embed_oembed_html( $cache, $url ) {
-		$parsed_url = wp_parse_url( $url );
-		if ( false === strpos( $parsed_url['host'], 'tumblr.com' ) ) {
-			return $cache;
+	public function sanitize_raw_embeds( Document $dom ) {
+		$nodes = $dom->xpath->query( sprintf( '//div[ @class and contains( concat( " ", normalize-space( @class ), " " ), " tumblr-post " ) and starts-with( @data-href, "%s" ) ]', $this->base_embed_url ) );
+
+		if ( 0 === $nodes->length ) {
+			return;
 		}
 
-		// @todo The iframe will not get sized properly.
-		if ( preg_match( '#data-href="(?P<href>https://embed.tumblr.com/embed/post/\w+/\w+)"#', $cache, $matches ) ) {
-			$cache = AMP_HTML_Utils::build_tag(
+		foreach ( $nodes as $node ) {
+			$iframe_src = $node->getAttribute( 'data-href' );
+
+			$attributes = [
+				'src'       => $iframe_src,
+				'layout'    => 'responsive',
+				'width'     => $this->args['width'],
+				'height'    => $this->args['height'],
+				'resizable' => '',
+				'sandbox'   => 'allow-scripts allow-popups allow-same-origin',
+			];
+
+			$amp_element = AMP_DOM_Utils::create_node(
+				$dom,
 				'amp-iframe',
+				$attributes
+			);
+
+			// Add an overflow element to allow the amp-iframe to be manually resized.
+			$overflow_element              = AMP_DOM_Utils::create_node(
+				$dom,
+				'button',
 				[
-					'width'   => $this->args['width'],
-					'height'  => $this->args['height'],
-					'layout'  => 'responsive',
-					'sandbox' => 'allow-scripts allow-popups', // The allow-scripts is needed to allow the iframe to render; allow-popups needed to allow clicking.
-					'src'     => $matches['href'],
-				],
-				sprintf( '<a placeholder href="%s">Tumblr</a>', $url )
+					'overflow' => '',
+					'style'    => 'bottom:0',
+				]
 			);
-		}
+			$overflow_element->textContent = __( 'See more', 'amp' );
+			$amp_element->appendChild( $overflow_element );
```

Comment:
@westonruter Would you consider adding styles to override the `:hover` and `:active` rules for the button also?

```suggestion

			$style = AMP_DOM_Utils::create_node( $dom, Tag::STYLE, [] );
			$style->textContent = 'amp-iframe > button:not(:hover):not(:active)[overflow], amp-iframe > button[overflow]:hover, amp-iframe > button[overflow]:active {
				color: black;
				background-color: white;
				border: 3px solid black;
			}';
			$overflow_element->appendChild( $style );

			$amp_element->appendChild( $overflow_element );
```

Reply:
Here's how it looks in Twenty-Twenty One:


https://user-images.githubusercontent.com/16200219/109265900-2b321380-77ff-11eb-9ffc-b3dd13466116.mp4

Reply:
OK, I checked the the themes other than Twenty Twenty-One and none of them have the problem, as they all have background colors set. Therefore, since it's exclusively a problem with Twenty Twenty-One. So how about adding adding the additional CSS via the core theme sanitizer? This one style rule seems all ot be needed:

```css
button[overflow]:hover {
	background-color: var(--global--color-background);
}
```

Reply:
Done in https://github.com/ampproject/amp-wp/pull/5926/commits/5c4b6676de0405b954323ca1e22e5c41843c9e2a.